src/core/insights/index.ts

```diff
@@ -0,0 +1,199 @@
+  posthogApiKey: string;
+  posthogApiHost: string;
+};
+
+let debugMode = false;
+
+export const initInsights = ({
+  mixpanelToken,
+  mixpanelAutoCapture,
+  posthogApiKey,
+  posthogApiHost,
+  debug = false,
+}: InsightsConfig) => {
+  debugMode = !!debug;
+
+  try {
+    mixpanel.initMixpanel(mixpanelToken, mixpanelAutoCapture, debugMode);
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to initialize Mixpanel", e);
+    }
+  }
+
+  try {
+    posthog.initPosthog(posthogApiKey, posthogApiHost);
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to initialize Posthog", e);
+    }
+  }
+};
+
+export const enableDebugMode = () => {
+  debugMode = true;
+  try {
+    mixpanel.enableDebugMode();
+    posthog.enableDebugMode();
+  } catch (e) {
+    console.error("Failed to enable debug mode", e);
+  }
+};
+
+export const disableDebugMode = () => {
+  debugMode = false;
+  try {
+    mixpanel.disableDebugMode();
+    posthog.disableDebugMode();
+  } catch (e) {
+    console.error("Failed to disable debug mode", e);
+  }
+};
+
+export const identify = ({
+  userId,
+  accountId,
+  organisationId,
+  email,
+  name,
+}: InsightsIdentity) => {
+  // In very rare cases we might have a user without an account, so we'll
+  // let null/undefined/blank strings through on that one
+  if (!userId) {
+    if (debugMode) {
+      console.warn("User ID not provided, skipping identify");
+    }
+    return;
+  }
+
+  try {
+    mixpanel.identify({ userId, accountId, organisationId, email, name });
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to identify user in Mixpanel", e);
+    }
+  }
+
+  try {
+    posthog.identify({ userId, accountId, organisationId, email, name });
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to identify user in Posthog", e);
+    }
+  }
+};
+
+export const trackPageView = () => {
+  try {
+    mixpanel.trackPageView();
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to track page view in Mixpanel", e);
+    }
+  }
+
+  try {
+    posthog.trackPageView();
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to track page view in Posthog", e);
+    }
+  }
+};
+
+export const track = (event: string, properties?: Record<string, unknown>) => {
+  try {
+    mixpanel.track(event, properties);
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to track event in Mixpanel", e);
+    }
+  }
+
+  try {
+    posthog.track(event, properties);
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to track event in Posthog", e);
+    }
+  }
+};
+
+type InsightAttributes = {
+  event?: string;
+  [key: string]: string | undefined;
+};
+
+export const setupObserver = () => {
+  // Helper to get all data-insight-* attributes from an element
+  const getInsightAttributes = (element): InsightAttributes => {
+    // limit how many data attributes we'll process
+    const MAX_ATTRIBUTES = 10;
```

Comment:
Why have this limit?

Reply:
I can see how this keeps the loop time down in extreme cases, but it seems like overkill when on the other hand it could be feasible in (admittedly rare) cases that we have a legit element with ten attributes and the insights tracker is no. 11 - all it takes is a bunch of arias to bloat it out.

Can keep it but I would increase the threshold a bit if so

Reply:
It only increments the count for `data-insights-*` attributes, not every single one. And if we have a case for sending an event with more than 9 properties (given one slot is the event name) then we could just switch to using `track()`